### PR TITLE
impl(generator): add rest stub generation

### PR DESCRIPTION
**Pull request by scotthart**
This adds rest stubs to the generator along with the associated golden files and test for the golden files.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209)

### CODE REVIEW

**Review by dbolduc — COMMENTED** [review]
2000 lines is an awfully big PR. I get that you have already written the unit tests and have a complete PR. But I worry that we will not get the most that we can out of the review process.

Is it too late to add a new golden proto and introduce the different types of RPCs one PR at a time?

😬

(I may be alone in this opinion, we can see how other team members feel. I will not be the one reviewing this code, so my opinion should be given little to no weight.)

File: generator/internal/stub_rest_generator.cc [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    "\n"
+    "class $stub_rest_class_name$ {\n"
+    " public:\n"
+    "  virtual ~$stub_rest_class_name$() = default;\n");
+  // clang-format on
+
+  for (auto const& method : methods()) {
+    HeaderPrintMethod(
+        method,
+        {MethodPattern(
```
Comment (coryan) [inline generator/internal/stub_rest_generator.cc:74]: I am going to insist that we stop adding these predicate-based functions and we use raw strings.  The code for things like the round robin decorator is much simpler to grok:

https://github.com/googleapis/google-cloud-cpp/blob/main/generator/internal/round_robin_decorator_generator.cc

File: generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+StatusOr<google::test::admin::database::v1::ListLogsResponse>
+DefaultGoldenKitchenSinkRestStub::ListLogs(
+      rest_internal::RestContext& rest_context,
+      google::test::admin::database::v1::ListLogsRequest const& request) {
+  return rest_internal::Get<google::test::admin::database::v1::ListLogsResponse>(
+      *rest_client_, rest_context, request,
+      absl::StrCat("/v2/",request.parent(),"/logs"),
+{std::make_pair("page_size", std::to_string(request.page_size())),
+std::make_pair("page_token", request.page_token())});
```
Comment (devbww) [inline generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc:75]: Can we take any reasonable measures to get the initial spacing and indentation right here (`method_rest_path` and `method_http_query_parameters`)?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209)* status: 0 of 16 files reviewed, 3 unresolved discussions (waiting on @coryan, @dbolduc, and @devbww)

___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGNRROf8ouqvHuXRK5V-r1-75:-NGORGygBWCVyg-TFvF1:bjmqbl8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eb47d353919a70cbbb1224ec153e7bed43e14d84/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc#L75)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Can we take any reasonable measures to get the initial spacing and indentation right here (`method_rest_path` and `method_http_query_parameters`)?
</blockquote></details>

I tried several things while implementing this and couldn't make the formatting work in all cases. I can create an issue to address this, but honestly I think the correct solution is to figure out how to emit clang-formatted code so there isn't this intermediate state that only exists in the golden files.
___
*[`generator/internal/stub_rest_generator.cc` line 74 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGNRROf8ouqvHuXRK5d-r1-74:-NGORu6L80hDN7FDfdox:bc8b3vj) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eb47d353919a70cbbb1224ec153e7bed43e14d84/generator/internal/stub_rest_generator.cc#L74)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I am going to insist that we stop adding these predicate-based functions and we use raw strings.  The code for things like the round robin decorator is much simpler to grok:

https://github.com/googleapis/google-cloud-cpp/blob/main/generator/internal/round_robin_decorator_generator.cc

</blockquote></details>

Code refactored.
___
*[`BUILD.bazel` line 309 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGNRROejMKczFap9MdF-r1-309:-NGORwJc9sE2VHzjeQ0K:ba3dfob) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eb47d353919a70cbbb1224ec153e7bed43e14d84/BUILD.bazel#L309)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I don't think these are necessary. Can't we just use the `//google/cloud:...` libs?
</blockquote></details>

Removed

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 11 of 16 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209)* status: all files reviewed, 10 unresolved discussions (waiting on @dbolduc, @devbww, and @scotthart)

___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.h` line 40 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOaxUPDzaVcTBbRsY9:-NGOaxUPDzaVcTBbRsYA:b-qyq558) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.h#L40)):*
> ```objc
> 
>   virtual StatusOr<google::test::admin::database::v1::GenerateAccessTokenResponse> GenerateAccessToken(
>       rest_internal::RestContext& rest_context,
> ```

Do we need to fully qualify `rest_internal::RestContext`? 
___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGNRROf8ouqvHuXRK5V-r1-75:-NGOhDHQEx0Jfmf6dlpG:bdsofkd) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eb47d353919a70cbbb1224ec153e7bed43e14d84/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc#L75)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I tried several things while implementing this and couldn't make the formatting work in all cases. I can create an issue to address this, but honestly I think the correct solution is to figure out how to emit clang-formatted code so there isn't this intermediate state that only exists in the golden files.
</blockquote></details>

Maybe something like #10213?  I am sure it is not perfect, but looks like it would be less strange?
___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc` line 33 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGObXdPCOVCxHgCU2io:-NGObXdQ4s0wyBeEzpbA:b-qp9mnr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc#L33)):*
> ```cpp
> DefaultGoldenKitchenSinkRestStub::DefaultGoldenKitchenSinkRestStub(Options options)
>     : rest_client_(rest_internal::MakePooledRestClient(
>           "https://goldenkitchensink.googleapis.com", options)),
> ```

How will applications override this endpoint?  It is Okay to postpone the fix for that, but if so, please open a bug.
___
*[`generator/integration_tests/golden/internal/golden_thing_admin_rest_stub.h` line 44 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOtA8K0_6y9y6vnANI:-NGOtA8LDRFvdTKie_sY:b-4yf6g4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/integration_tests/golden/internal/golden_thing_admin_rest_stub.h#L44)):*
> ```objc
>       google::test::admin::database::v1::ListDatabasesRequest const& request) = 0;
> 
>   virtual StatusOr<google::longrunning::Operation> CreateDatabase(
> ```

I am missing the functions to poll these operations.  How will those happen?  It is Okay if this is planned for a future PR.
___
*[`generator/integration_tests/tests/golden_kitchen_sink_rest_stub_test.cc` line 47 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOcBlEC87-Z0_mmGk-:-NGOcBlEC87-Z0_mmGk0:b-unr7s6) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/integration_tests/tests/golden_kitchen_sink_rest_stub_test.cc#L47)):*
> ```cpp
> auto constexpr kServiceUnavailable = "503 Service Unavailable";
> 
> class GoldenKitchenSinkRestStubTest : public ::testing::Test {
> ```

Are you getting enough value out of this fixture?  Seems like all you get out of it is removing a single line in each test:

```cc
auto mock_rest_client = absl::make_unique<MockRestClient>();
```

The downside is that one has to go back several pages to understand what `mock_rest_client_` and `service_unavailable_` happen to be.

Consider just removing the fixture.
___
*[`generator/internal/stub_rest_generator.cc` line 74 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGNRROf8ouqvHuXRK5d-r1-74:-NGOdwto4MPnC4s85DRa:btzuspr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eb47d353919a70cbbb1224ec153e7bed43e14d84/generator/internal/stub_rest_generator.cc#L74)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Code refactored.
</blockquote></details>

Consider this:

```cc
  for(auto const& method : methods()) {
    if (IsStreaming(method) || !HasHttpAnnotation(method)) continue;
    if (IsResponseTypeEmpty(method)) {
      HeaderPrintMethod(method, __FILE__, __LINE__, R"""(
  virtual Status $method_name(
      rest_internal::RestContext& rest_context,
      $request_type$ const& request) = 0;
)""");
      continue;
    }
    HeaderPrintMethod(method, __FILE__, __LINE__, R"""(
  virtual StatusOr<$response_type$> $method_name(
      rest_internal::RestContext& rest_context,
      $request_type$ const& request) = 0;
)""");
  }
```

Yes, it duplicates some code, but it puts all the code for each method in a single raw string that is easy to scan / read.
___
*[`generator/internal/stub_rest_generator.cc` line 38 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOfFNYC7ShKRhzgWue:-NGOfFNYC7ShKRhzgWuf:b-qszt35) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/internal/stub_rest_generator.cc#L38)):*
> ```cpp
> Status StubRestGenerator::GenerateHeader() {
>   HeaderPrint(CopyrightLicenseFileHeader());
>   HeaderPrint(  // clang-format off
> ```

These blocks of text can be represented by raw strings that are easier to read and edit.
___
*[`generator/internal/stub_rest_generator.cc` line 91 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOezgR5tSUL_M5y6vJ:-NGOezgR5tSUL_M5y6vK:b-236lpi) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/internal/stub_rest_generator.cc#L91)):*
> ```cpp
> 
>   // default stub class
>   HeaderPrint(  // clang-format off
> ```

These two blocks also can go into a single raw string and you can avoid all the `// clang-format off` stuff.
___
*[`generator/internal/stub_rest_generator.cc` line 106 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOgEIeBAcfRH2LSenc:-NGOgEIeBAcfRH2LSend:bqux2j0) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/internal/stub_rest_generator.cc#L106)):*
> ```cpp
> 
>   for (auto const& method : methods()) {
>     if (HasHttpAnnotation(method) && IsNonStreaming(method)) {
> ```

same comments as above, a bit of text duplication can increase readability here.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209)* status: all files reviewed, 10 unresolved discussions (waiting on @coryan, @dbolduc, and @devbww)

___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.h` line 40 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOaxUPDzaVcTBbRsY9:-NGSNfMIAyT1h6yfFl-2:b-1sfdtz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.h#L40)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Do we need to fully qualify `rest_internal::RestContext`? 
</blockquote></details>

It would probably be prudent to do so. Fixed.
___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc` line 33 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGObXdPCOVCxHgCU2io:-NGSM4DIBklFkQmON5MK:b659nv) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc#L33)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

How will applications override this endpoint?  It is Okay to postpone the fix for that, but if so, please open a bug.
</blockquote></details>

Created an issue to track.
___
*[`generator/integration_tests/golden/internal/golden_thing_admin_rest_stub.h` line 44 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOtA8K0_6y9y6vnANI:-NGSMAPsAa_EpzJ2DtbT:b-pbcgm0) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/integration_tests/golden/internal/golden_thing_admin_rest_stub.h#L44)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I am missing the functions to poll these operations.  How will those happen?  It is Okay if this is planned for a future PR.
</blockquote></details>

Future PR. I'm still trying to figure out the best way to design/implement this without a CompletionQueue. 
___
*[`generator/integration_tests/tests/golden_kitchen_sink_rest_stub_test.cc` line 47 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOcBlEC87-Z0_mmGk-:-NGSWPcw98BP5inP3yaE:b-rthuc6) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/integration_tests/tests/golden_kitchen_sink_rest_stub_test.cc#L47)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Are you getting enough value out of this fixture?  Seems like all you get out of it is removing a single line in each test:

```cc
auto mock_rest_client = absl::make_unique<MockRestClient>();
```

The downside is that one has to go back several pages to understand what `mock_rest_client_` and `service_unavailable_` happen to be.

Consider just removing the fixture.
</blockquote></details>

Fixture removed.
___
*[`generator/internal/stub_rest_generator.cc` line 74 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGNRROf8ouqvHuXRK5d-r1-74:-NGSfMB87HxsM33Gt3f7:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eb47d353919a70cbbb1224ec153e7bed43e14d84/generator/internal/stub_rest_generator.cc#L74)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Consider this:

```cc
  for(auto const& method : methods()) {
    if (IsStreaming(method) || !HasHttpAnnotation(method)) continue;
    if (IsResponseTypeEmpty(method)) {
      HeaderPrintMethod(method, __FILE__, __LINE__, R"""(
  virtual Status $method_name(
      rest_internal::RestContext& rest_context,
      $request_type$ const& request) = 0;
)""");
      continue;
    }
    HeaderPrintMethod(method, __FILE__, __LINE__, R"""(
  virtual StatusOr<$response_type$> $method_name(
      rest_internal::RestContext& rest_context,
      $request_type$ const& request) = 0;
)""");
  }
```

Yes, it duplicates some code, but it puts all the code for each method in a single raw string that is easy to scan / read.


</blockquote></details>

Done.
___
*[`generator/internal/stub_rest_generator.cc` line 38 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOfFNYC7ShKRhzgWue:-NGSbu9P1-XZfed_S79-:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/internal/stub_rest_generator.cc#L38)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

These blocks of text can be represented by raw strings that are easier to read and edit.
</blockquote></details>

Done.
___
*[`generator/internal/stub_rest_generator.cc` line 91 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOezgR5tSUL_M5y6vJ:-NGSbwui0Y5AHZcm5jwZ:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/internal/stub_rest_generator.cc#L91)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

These two blocks also can go into a single raw string and you can avoid all the `// clang-format off` stuff.
</blockquote></details>

Done.
___
*[`generator/internal/stub_rest_generator.cc` line 106 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGOgEIeBAcfRH2LSenc:-NGSfSki9D76bHkeHvbJ:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/edb9d8a1e69f4ef28e10589b8e5cf265c6f1e654/generator/internal/stub_rest_generator.cc#L106)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

same comments as above, a bit of text duplication can increase readability here.
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 7 of 7 files at r3, 1 of 1 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209)* status: all files reviewed, 2 unresolved discussions (waiting on @dbolduc and @devbww)

___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGNRROf8ouqvHuXRK5V-r1-75:-NGShYtX9wSUZF87yv2k:beqb0wr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eb47d353919a70cbbb1224ec153e7bed43e14d84/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc#L75)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Maybe something like #10213?  I am sure it is not perfect, but looks like it would be less strange?
</blockquote></details>

I assume you are not opposed to that change?

**Review by coryan — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-:-NGY2xap-r2ueQtbveJ2:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @dbolduc and @devbww)

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209)* status: all files reviewed, 2 unresolved discussions (waiting on @dbolduc and @devbww)

___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10209#-NGNRROf8ouqvHuXRK5V-r1-75:-NGY35L3EM7oijQ_zONe:b-xvjtnr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eb47d353919a70cbbb1224ec153e7bed43e14d84/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc#L75)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I assume you are not opposed to that change?
</blockquote></details>

I'm fine with it.

**Pull request merged**